allow user to pass in custom partitioner function

xuanzhang816 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157580



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

xuanzhang816 committed:
allow user to pass in custom partitioner function

[ghstack-poisoned]

xuanzhang816 commented:
@xuanzhang816 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77751433).

xuanzhang816 commented:
@xuanzhang816 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D77753038).

xuanzhang816 commented:
@pytorchbot label "topic: not user facing"

mlazos commented:
Should you add a test for this?

bdhirsh commented:
also cc @zou3519, this is the "custom partitioner" we talked about

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -2295,10 +2295,17 @@ def partition_fn(
                 "static_lifetime_input_indices", None
             )
 
+            # TODO(xuanzh): do I need to change the first argument of dynamo_timed below?
```
yeah, i think it would be nice if customize the dynamo_timed message below to print something like f`custom_partitioner_pass: {type(custom_partition_fn)}` so it's a bit easier to tell from tlparse if or what custom partitioner someone is using in their job

xuanzhang816 committed:
Update on "allow user to pass in custom partitioner function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77753038](https://our.internmc.facebook.com/intern/diff/D77753038)

[ghstack-poisoned]

xuanzhang816 commented:
@bdhirsh I did a rewrite of this to ensure cacheability, copying the idea of custom graph pass, and to add tests.

xuanzhang816 commented on torch/_inductor/config.py:
```diff
@@ -261,6 +261,11 @@ def prologue_fusion_enabled() -> bool:
 
+# Allow users to pass in custom partition function
+custom_partition_fn: Optional[
+    Callable[..., tuple[torch.fx.GraphModule, torch.fx.GraphModule]]
```
added.

xuanzhang816 committed:
Update on "allow user to pass in custom partitioner function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D77753038](https://our.internmc.facebook.com/intern/diff/D77753038)

[ghstack-poisoned]

ezyang commented:
Let's wait for Brian then

xuanzhang816 committed:
Update

[ghstack-poisoned]

xuanzhang816 committed:
Update

[ghstack-poisoned]

xuanzhang816 committed:
Update

[ghstack-poisoned]

xuanzhang816 committed:
Update

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/custom_partitioner_fn.py:
```diff
@@ -0,0 +1,54 @@
+import torch
+
+
+class CustomPartitionerFn(ABC):
```
nit: I would probably bundle this into the same file that the other `CustomGraphPass` object definitions live (`custom_graph_pass.py`), since at the end of the day the purpose of the extension is the same (custom graph pass accepting (graph, args)), just with a slightly different signature

bdhirsh commented on torch/_inductor/compile_fx.py:
```diff
@@ -2110,16 +2111,30 @@ def partition_fn(
+    else:
+        assert isinstance(config.custom_partitioner_fn, CustomPartitionerFn)
+        with dynamo_utils.dynamo_timed(
+            config.custom_partitioner_fn.__class__.__name__,
```
nice, having tlparse tell us if "slow partitioner compile times" are actually coming from a user-provided partitioner will make debugging a lot easier.

bdhirsh commented on test/inductor/test_custom_partitioner_fn.py:
```diff
@@ -0,0 +1,71 @@
+# Owner(s): ["module: inductor"]
```
nit: this can be `"module: pt2-dispatcher"` (since partitioner is owned by pt2-dispatcher, although I agree it's confusing since inductor is the one that calls it 😛 )

xuanzhang816 committed:
Update

[ghstack-poisoned]

xuanzhang816 committed:
Update

[ghstack-poisoned]

bdhirsh reviewed (approved):
thanks for updating!

xuanzhang816 committed:
Update

[ghstack-poisoned]
